### PR TITLE
Adding component for classifier with tf.estimator

### PR DESCRIPTION
**Pull request by weizhenzhao**
tf.estimator is a higher level api , which is easy to support larger scale distribute trainning and even can change classifier method without modify much code .

**Proposed changes**: integrate with tf.estimator to support tensorflow higher api even google's autoML in the future to improve the classificaiton accuracy .
- ...

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [Y ] added some tests for the functionality
- [x] updated the documentation
- [Y] updated the changelog

below is an training log and prediction log :
[trainning_and_predict_log.txt.txt](https://github.com/GaoQ1/rasa_nlu_gq/files/3000192/trainning_and_predict_log.txt.txt)

image

mainly change code  in training process

`        num_classes = len(intent_dict)

        head = tf.contrib.estimator.multi_class_head(n_classes=num_classes)

        feature_names = ['a_in']
        self.feature_columns = [tf.feature_column.numeric_column(key=k,shape=[1, X.shape[1]]) for k in feature_names]

        x_tensor = {'a_in': X}
        intents_for_X = intents_for_X.astype(np.int32)

        tf.logging.set_verbosity(tf.logging.INFO)
        config_proto = self.get_config_proto(self.component_config)

        #sparse_softmax_cross_entropy

        self.estimator = tf.contrib.estimator.LinearEstimator(
                                                     head = head,
                                                     feature_columns=self.feature_columns,
                                                     optimizer='Ftrl',
                                                     config=tf.estimator.RunConfig(session_config=config_proto)
                                                 )
        self.estimator.train(input_fn=lambda: self.input_fn(x_tensor,
                                                  intents_for_X,
                                                  self.batch_size,
                                                  shuffle_num=1000,
                                                  mode = tf.estimator.ModeKeys.TRAIN),
                                                  max_steps=2000)
        results = self.estimator.evaluate(input_fn=lambda: self.input_fn(x_tensor,
                                                  intents_for_X,
                                                  self.batch_size,
                                                  shuffle_num=1000,
                                                  mode = tf.estimator.ModeKeys.PREDICT))

        print(results)`

mainly code change in load process 

`

predict = predictor.from_saved_model(export_dir=os.path.join(model_dir,file_name),config=config_proto)

`

and the prediction process 

`

X = message.get("text_features").tolist()
            examples = []
            feature = {}
            feature['a_in'] = tf.train.Feature(float_list=tf.train.FloatList(value=X))
            example = tf.train.Example(
                features=tf.train.Features(
                    feature=feature
                )
            )
            examples.append(example.SerializeToString())

            # Make predictions.
            result_dict = self.predictor({'inputs': examples})
            result_score_list = result_dict['scores'][0]
            max_score = np.max(result_dict['scores'][0])
            max_index = np.argmax(result_dict['scores'][0])

`

and the persist process 

` 

feature_spec = tf.feature_column.make_parse_example_spec(self.feature_columns)
        serving_input_receiver_fn = tf.estimator.export.build_parsing_serving_input_receiver_fn(feature_spec)
        path = self.estimator.export_savedmodel(model_dir, serving_input_receiver_fn)
        file_dir = os.path.basename(path).decode('utf-8')

`

new feature of tensorflow used 

tf.data : an load and extract data in high profermance
tf.estimator : tf higher level api , which can support large scale distribute ml and easy to convert to 
                      another method without need to change much code
tf.saved_model : it can export model to *.pb format which can be loaded in other frameworks like 
                          pytorch , keras and other ml frameworks , also support tensorflow serving

thanks
weizhen

### CODE REVIEW

File: rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+        # visualization of accuracy
+        "evaluate_every_num_epochs": 10,  # small values may hurt performance
+        "evaluate_on_num_examples": 1000,  # large values may hurt performance
+
+        "config_proto": {
+            "device_count": cpu_count(),
+            "inter_op_parallelism_threads": 0,
+            "intra_op_parallelism_threads": 0,
+            "allow_growth": True,
+            "allocator_type": 'BFC',
```
Comment (GaoQ1) [inline rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py]: 这个参数可以加个注释

Comment (GaoQ1) [reply rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py]: BFC means a "Best-fit with coalescing" algorithm

File: rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+            "per_process_gpu_memory_fraction": 0.5
```
Comment (GaoQ1) [inline rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py]: 这个参数为啥给定0.5

Comment (weizhenzhao) [reply rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py]: 这个0.5是因为那个腾讯的bert在这边电脑上面最高可能占一半的GPU显存，然后为了不影响腾讯那个所以把这个设置为0.5

File: rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+        self.hidden_layer_size = self.component_config['hidden_layer_size']
+
+        self.batch_size = self.component_config['batch_size']
+        self.epochs = self.component_config['epochs']
+        self.learning_rate = self.component_config['learning_rate']
+
+    def _load_regularization_params(self):
+        self.C2 = self.component_config['C2']
+        self.droprate = self.component_config['droprate']
+
```
Comment (GaoQ1) [inline rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py]: 正则化这块你好像下面并没有用到，self.C2, self.droprate

Comment (weizhenzhao) [reply rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py]: 好的 我等把这块给删了

File: rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+        meta = model_metadata.for_component(cls.name)
+        config_proto = cls.get_config_proto(meta)
+
+        print("bert model loaded")
+
+        if model_dir and meta.get("classifier_file"):
+            file_name = meta.get("classifier_file")
+
+            predict = predictor.from_saved_model(export_dir=os.path.join(model_dir,file_name),config=config_proto)
```
Comment (GaoQ1) [inline rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py]: predictor 这个参数是在开始`from tensorflow.contrib import predictor`这儿引入的，但是和`self.predictor = preditor`容易混淆，最好将这个换个名字。比如`from tensorflow.contrib import predictor as Pred`或者在`def load()`方法下面再引入`from tensorflow.contrib import predictor`

File: rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py (hunk 2 above)
Comment (GaoQ1) [reply rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py]: 这块的话加个注释吧

File: rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py (hunk 4 above)
Comment (weizhenzhao) [reply rasa_nlu_gao/classifiers/embedding_bert_intent_estimator_classifier.py]: 好的 ， 我改一下

**Comment by GaoQ1** [comment]
其他的没啥问题，还有你可以修改下README加上你这个component的pipeline配置例子。同样最好的话你可以写个简单的blog介绍，然后可以将链接加到[rasa_chatbot_cn](https://github.com/GaoQ1/rasa_chatbot_cn)这个README上  : )

**Comment by weizhenzhao** [comment]
> @weizhenzhao 其他的没啥问题，还有你可以修改下README加上你这个component的pipeline配置例子。同样最好的话你可以写个简单的blog介绍，然后可以将链接加到[rasa_chatbot_cn](https://github.com/GaoQ1/rasa_chatbot_cn)这个README上 : )

好的 ， 好的

**Comment by GaoQ1** [comment]
你好像新增了好多venv/Script文件，这个是你的环境配置的东西吧，这个不要提交。或者你在.gitignore上加上。
